content/en/blog/_posts/2025-01-22-kube-scheduler-simulator/index.md

```diff
@@ -0,0 +1,138 @@
+    kube-scheduler-simulator.sigs.k8s.io/score-result: >-
+      {"node-jjfg5":{"ImageLocality":"0","NodeAffinity":"0","NodeResourcesBalancedAllocation":"52","NodeResourcesFit":"47","TaintToleration":"0","VolumeBinding":"0"},"node-mtb5x":{"ImageLocality":"0","NodeAffinity":"0","NodeResourcesBalancedAllocation":"76","NodeResourcesFit":"73","TaintToleration":"0","VolumeBinding":"0"}}
+    kube-scheduler-simulator.sigs.k8s.io/selected-node: node-mtb5x
+```
+
+Users can also integrate their custom plugins or extenders, into the debuggable scheduler and visualize their results. 
+
+This debuggable scheduler can also run standalone, e.g., on any Kubernetes cluster or in integration tests. 
+This would be useful to custom plugin developers who want to test their plugins or examine their custom scheduler in a real cluster with better debuggability.
+
+## The simulator as a better dev cluster
+
+As mentioned earlier, with a limited set of tests, it is impossible to predict every possible scenario in a real-world cluster.
+Typically, you will test the scheduler in a small, development cluster before deploying it to production, hoping that no issues arise.
+
+[The simulator’s importing feature](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/import-cluster-resources.md)
+provides a solution by allowing you to simulate deploying a new scheduler version in a production-like environment without impacting your live workloads.
+
+By continuously syncing between your production cluster and the simulator, you can safely test a new scheduler version with the same resources your production cluster handles. 
+Once confident in its performance, you can proceed with the production deployment, reducing the risk of unexpected issues.
+
+## What are the use cases?
+
+1. **Cluster users**: Examine if your scheduling constraints (e.g., PodAffinity, PodTopologySpread) work as intended.
+1. **Cluster admins**: Assess how your cluster would behave with changes to the scheduler configuration.
+1. **Scheduler plugin developers**: Test your custom scheduler plugins or extenders, use the debuggable scheduler in integration tests or development clusters, or use the [syncing](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/simulator/v0.3.0/simulator/docs/import-cluster-resources.md) feature for testing within a production-like environment.
+
+## Getting started
+
+The simulator only requires Docker to be installed on your machine; a Kubernetes cluster is not necessary.
```

Comment:
how will it access the cluster for the use cases listed above? This is very confusing for me, I do not know what to expect when I run this. If it doesn't need a cluster, what scheduler version will it even use? Which feature gates enabled? Is it compatible with [my cloud] scheduler?

Reply:
> how will it access the cluster for the use cases listed above?

It internally runs kwok cluster which you can access.

Regarding other points, we don't have compatible matrics or the feature gate control yet, which we have to improve.